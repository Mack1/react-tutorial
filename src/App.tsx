import React from 'react';
import axios from 'axios';
import { sortBy } from 'lodash';

const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';


const useSemiPersistentState = (
  key: string,
  initialState: string
): [string, (newValue: string) => void] => {
  const [value, setValue] = React.useState(
    localStorage.getItem(key) || initialState
  );
  React.useEffect(() => {
    localStorage.setItem(key, value);
  }, [value, key]);
  return [value, setValue];
};


type Story = {
  objectID: string;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
};

type Stories = Array<Story>;

type StoriesState = {
  data: Stories;
  isLoading: boolean;
  isError: boolean;
};

interface StoriesFetchInitAction {
  type: 'STORIES_FETCH_INIT';
}

interface StoriesFetchSuccessAction {
  type: 'STORIES_FETCH_SUCCESS';
  payload: Stories;
}

interface StoriesFetchFailureAction {
  type: 'STORIES_FETCH_FAILURE';
}

interface StoriesRemoveAction {
  type: 'REMOVE_STORY';
  payload: Story;
}

type StoriesAction =
  | StoriesFetchInitAction
  | StoriesFetchSuccessAction
  | StoriesFetchFailureAction
  | StoriesRemoveAction;

const storiesReducer = (
  state: StoriesState,
  action: StoriesAction
) => {
  switch (action.type) {
    case 'STORIES_FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case 'STORIES_FETCH_FAILURE':
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case 'REMOVE_STORY':
      return {
        ...state,
        data: state.data.filter(
          story => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};
const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState(
    'search',
    'React'
  );
  const [url, setUrl] = React.useState(
    `${API_ENDPOINT}${searchTerm}`
  );
  const [stories, dispatchStories] = React.useReducer(
    storiesReducer,
    { data: [], isLoading: false, isError: false }
  );
  const handleFetchStories = React.useCallback(async () => {
    dispatchStories({ type: 'STORIES_FETCH_INIT' });
    try {
      const result = await axios.get(url);
      dispatchStories({
        type: 'STORIES_FETCH_SUCCESS',
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: 'STORIES_FETCH_FAILURE' });
    }
  }, [url]);
  React.useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);


  const handleRemoveStory = (item: Story) => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item,
    });
  };


  const handleSearchInput = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    setSearchTerm(event.target.value);
  };


  const handleSearchSubmit = (
    event: React.FormEvent<HTMLFormElement>
  ) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);

    event.preventDefault();
  };
  return (
    <div>
      <h1>My Hacker Stories</h1>
      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />
      <hr />
      {stories.isError && <p>Something went wrong ...</p>}
      {stories.isLoading ? (
        <p>Loading ...</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </div>
  );
};

type SearchFormProps = {
  searchTerm: string;
  onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onSearchSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
};

const SearchForm = ({
  searchTerm,
  onSearchInput,
  onSearchSubmit,
}: SearchFormProps) => (
  <form onSubmit={onSearchSubmit}>
    <InputWithLabel
      id="search"
      value={searchTerm}
      isFocused
      onInputChange={onSearchInput}
    >
      <strong>Search:</strong>
    </InputWithLabel>
    <button type="submit" disabled={!searchTerm}>
      Submit
    </button>
  </form>
);

type InputWithLabelProps = {
  id: string;
  value: string;
  type?: string;
  onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  isFocused?: boolean;
  children: React.ReactNode;
};

const InputWithLabel = ({
  id,
  value,
  type = 'text',
  onInputChange,
  isFocused,
  children,
}: InputWithLabelProps) => {
  const inputRef = React.useRef<HTMLInputElement>(null!);

  React.useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);
  return (
    <>
      <label htmlFor={id}>{children}</label>
      &nbsp;
      <input
        ref={inputRef}
        id={id}
        type={type}
        value={value}
        onChange={onInputChange}
      />
    </>
  );
};

type list = {
  list: Array<object>;
};

type sort = { sort: object };

const SORTS:any = {
  NONE: ({ list }: list) => list,
  TITLE: ({ list }: list) => sortBy(list, 'title'),
  AUTHOR: ({ list }: list) => sortBy(list, 'author'),
  COMMENT: ({ list }: list) => sortBy(list, 'num_comments').reverse(),
  POINT: ({ list }: list) => sortBy(list, 'points').reverse(),
};

type ListProps = {
  list: Stories;
  onRemoveItem: (item: Story) => void;
};

type ItemProps = {
  objectID: string;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
  item: Story;
  onRemoveItem: (item: Story) => void;
};

const List = ({ list, onRemoveItem }: ListProps) => {
  const [sort, setSort] = React.useState('NONE');

  type KeyProps = {
    sortKey: string;
  };

  const handleSort = ({ sortKey }: KeyProps) => {
    setSort(sortKey); 
  };

  const sortFunction = SORTS[sort];
  const sortedList = sortFunction(list);

  return (
    <div>
      <div style={{ display: 'flex' }}>
        <span style={{ width: '40%' }}>
          <button type="button" onClick={() => handleSort({ sortKey: 'TITLE' })}>
            Title
          </button>
        </span>
        <span style={{ width: '30%' }}>Author</span>
          <button type="button" onClick={() => handleSort({ sortKey: 'AUTHOR' })}>
            Author
          </button>
        <span style={{ width: '10%' }}>Comments</span>
          <button type="button" onClick={() => handleSort({ sortKey: 'COMMENT' })}>
            Comments
          </button>
        <span style={{ width: '10%' }}>Points</span>
          <button type="button" onClick={() => handleSort({ sortKey: 'POINT' })}>
            Points
          </button>
        <span style={{ width: '10%' }}>Actions</span>
      </div>

      {sortedList.map( (item: ItemProps) => (
        <Item
          key={item.objectID}
          item={item}
          onRemoveItem={onRemoveItem}
        />
      ))}
    </div>
   );
};

const Item = ({ Story, onRemoveItem }: ItemProps) => (
  <div style={{ display: 'flex'}}>
    <span style={{ width: '40%'}}>
      <a href={Story.url}>{Story.title}</a>
    </span>
    <span style={{ width: '30%'}}>{Story.author}</span>
    <span style={{ width: '10%'}}>{Story.num_comments}</span>
    <span style={{ width: '10%'}}>{Story.points}</span>
    <span style={{ width: '10%'}}>
      <button type="button" onClick={() => onRemoveItem(Story)}>
        Dismiss
      </button>
    </span>
  </div>
);
export default App;

export { storiesReducer, SearchForm, InputWithLabel, List, Item };