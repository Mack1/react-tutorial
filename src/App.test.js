import React from 'react';
import App, {
	storiesReducer,
	Item,
	List,
	SearchForm,
	InputWithLabel,
} from './App';

import {
  render,
  screen,
  fireEvent,
  act,
} from '@testing-library/react';

import axios from 'axios';

jest.mock('axios');

//test('renders learn react link', () => {
//  const { getByText } = render(<App />);
//  const linkElement = getByText(/learn react/i);
//  expect(linkElement).toBeInTheDocument();
//});

describe('something thruthy and falsy', () => {
	it('true to be true', ()=> {
		expect(true).toBe(true);
	});

	it('false to be false', ()=> {
		expect(false).toBe(false);
	});	
});

test('false to be false', () => {
	expect(false).toBe(false);
});

describe('App component', () => {
	test('removes an item when clicking the Dismiss button', () => {

	});

	test('requests some initial stories from an API', () => {

	});
});

describe('something truthy and falsy', () => {
	test('true to be true', () => {
		expect(true).toBeTruthy();
	});

	test('false to be false', () => {
		expect(false).toBeFalsy();
	});
});

const storyOne = {
	title: 'React',
	url: 'https://reactjs.org/',
	author: 'Jordan Walke',
	num_comments: 3,
	points: 4,
	objectID: 0,
};

const storyTwo = {
	title: 'Redux',
	url: 'https://redux.js.org/',
	author: 'Dan Abramov, Andrew Clark',
	num_comments: 2,
	points: 5,
	objectID: 1,
};

const stories = [storyOne, storyTwo];

describe('storiesReducer', () => {
	test('removes a story from all stories', () => {
		const action = { type: 'REMOVE_STORY', payload: storyOne};
		const state = { data: stories, isLoading: false, isError: false};

		const newState = storiesReducer(state, action);

		const expectedState = {
			data: [storyTwo],
			isLoading: false,
			isError: false,
		};
		expect(newState).toStrictEqual(expectedState);
	});
});

describe('storiesReducer', () => {
	test('stories failed to fetch', () => {
		const action = { type: 'STORIES_FETCH_FAILURE'};
		const state = { data: stories, isLoading: false, isError: true};

		const newState = storiesReducer(state, action);

		const expectedState = {
			data: [storyOne, storyTwo],
			isLoading: false,
			isError: true,
		};
		expect(newState).toEqual(expectedState);
	});
});

describe('storiesReducer', () => {
	test('stories fetch success', () => {
		const action = { type: 'STORIES_FETCH_SUCCESS', payload: stories};
		const state = { data: stories, isLoading: false, isError: false};

		const newState = storiesReducer(state, action);

		const expectedState = {
			data: [storyOne, storyTwo],
			isLoading: false,
			isError: false,
		};
		expect(newState).toEqual(expectedState);
	});
});

describe('storiesReducer', () => {
	test('stories fetch init', () => {
		const action = { type: 'STORIES_FETCH_INIT'};
		const state = { data: stories, isLoading: true, isError: false};

		const newState = storiesReducer(state, action);

		const expectedState = {
			data: [storyOne, storyTwo],
			isLoading: true,
			isError: false,
		};
		expect(newState).toEqual(expectedState);
	});
});

// Testing the Item component, assert whether it renders expected HTML out from given props.
describe('Item', () => {
	test('renders all properties', () => {
		render(<Item item={storyOne} />);

		expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
		expect(screen.getByText('React')).toHaveAttribute(
			'href',
			'https://reactjs.org/'
			);

		screen.debug();
	});

	test('renders a clickable dismiss button', () => {
		render(<Item item={storyOne} />);

		expect(screen.getByRole('button')).toBeInTheDocument();
	});

	test('clicking the dismiss button calls the callback handler', () => {
		const handleRemoveItem = jest.fn();

		render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />);

		fireEvent.click(screen.getByRole('button'));

		expect(handleRemoveItem).toHaveBeenCalledTimes(1);
	});
});

//Testing the InPutWithLabel component that's a part of SearchForm
describe('SearchForm', () => {
	const searchFormProps = {
		searchTerm: 'React',
		onSearchInput: jest.fn(),
		onSearchSubmit: jest.fn(),
	};

	test('renders the input field with its value', () => {
		render(<SearchForm {...searchFormProps} />);

		expect(screen.getByDisplayValue('React')).toBeInTheDocument();
	});

	test('renders the correct label', () => {
		render(<SearchForm {...searchFormProps} />);

		expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
	});

	test('calls onSearchInput on input field change', () => {
		render(<SearchForm {...searchFormProps} />);

		fireEvent.change(screen.getByDisplayValue('React'), {
			target: { value: 'Redux' },
		});

		expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
	});

	test('calls onSearchSubmit on button submit click', () => {
		render(<SearchForm {...searchFormProps} />);

		fireEvent.submit(screen.getByRole('button'));

		expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
	});

	//Snapshot testing
	
	test('renders snapshot of SearchForm', () => {
		const { container } = render(<SearchForm {...searchFormProps} />);
		expect(container.firstChild).toMatchSnapshot();
	});

});


//Testing list component
describe('List', () => {
	const listProps = {
		list: stories,
		onRemoveItem: jest.fn(),
	};

	test('renders the list', () => {
		render(<List {...listProps} />);

		expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
		expect(screen.getByText('React')).toHaveAttribute(
			'href',
			'https://reactjs.org/'
		);
		//screen.debug();
	});

	test('renders snapshot of List', () => {
		const { container } = render(<List {...listProps} />);
		expect(container.firstChild).toMatchSnapshot();
	});

});

//Testing InputWithLabel component
/*
describe('InputWithLabel', () => {
	const labelProps = {
	  id: 1,
	  value: 'value',
	  type: 'text',
	  onInputChange: jest.fn(),
	  isFocused: true,
	};

	test('renders the InputWithLabel', () => {
		render(<InputWithLabel {...labelProps} />);

		expect(screen.getByText("")).toBeInTheDocument();
	});
});
*/


//Integration Testing App Component
describe('App', () => {
	test('succeeds fetching data', async () => {
		const promise = Promise.resolve({
			data: {
				hits: stories,
			},
		});

		axios.get.mockImplementationOnce(() => promise);

		render(<App />);

		expect(screen.queryByText(/Loading/)).toBeInTheDocument();

		await act(() => promise);

		expect(screen.queryByText(/Loading/)).toBeNull();

		expect(screen.getByText('React')).toBeInTheDocument();
		expect(screen.getByText('Redux')).toBeInTheDocument();
		expect(screen.getAllByText('Dismiss').length).toBe(2);
	});

	test('fails fetching data', async () => {
		const promise = Promise.reject();

		axios.get.mockImplementationOnce(() => promise);

		render(<App />);

		expect(screen.getByText(/Loading/)).toBeInTheDocument();

		try {
			await act(() => promise);
		} catch (error) {
			expect(screen.queryByText(/Loading/)).toBeNull();
			expect(screen.queryByText(/went wrong/)).toBeInTheDocument();
		}
	});

	test('removes a story', async () => {
		const promise = Promise.resolve({
			data: {
				hits: stories,
			},
		});

		axios.get.mockImplementationOnce(() => promise);

		render(<App />);

		await act(() => promise);

		expect(screen.getAllByText('Dismiss').length).toBe(2);
		expect(screen.getByText('Jordan Walke')).toBeInTheDocument();

		fireEvent.click(screen.getAllByText('Dismiss')[0]);

		expect(screen.getAllByText('Dismiss').length).toBe(1);
		expect(screen.queryByText('Jordan Walke')).toBeNull();
	});

	test('searches for specific stories', async () => {
		const reactPromise = Promise.resolve({
			data: {
				hits: stories,
			}
		});

		const anotherStory = {
			title: 'Javascript',
			url: 'https://en.wikipedia.org/wiki/JavaScript',
			author: 'Brendan Eich',
			num_comments: 15,
			points: 10,
			objectId: 3,
		};

		const javascriptPromise = Promise.resolve({
			data: {
				hits: [anotherStory],
			},
		});

		axios.get.mockImplementation(url => {
			if (url.includes('React')) {
				return reactPromise;
			}
			if (url.includes('JavaScript')) {
				return javascriptPromise;
			}

			throw Error();
		});

		// Initial Render

		render(<App />);

		//First Data Fetching

		await act(() => reactPromise);

		expect(screen.queryByDisplayValue('React')).toBeInTheDocument();
		expect(screen.queryByDisplayValue('JavaScript')).toBeNull();

		expect(screen.queryByText('Jordan Walke')).toBeInTheDocument();

		expect(
				screen.queryByText('Dan Abramov, Andrew Clark')
			).toBeInTheDocument();
		expect(screen.queryByText('Brendan Eich')).toBeNull();

		//User Interaction -> Search

		fireEvent.change(screen.queryByDisplayValue('React'), {
			target: {
				value: 'JavaScript',
			}
		});

		expect(screen.queryByDisplayValue('React')).toBeNull();
		expect(
			screen.queryByDisplayValue('JavaScript')
			).toBeInTheDocument();

		fireEvent.submit(screen.queryByText('Submit'));

		// Second Data Fetching

		await act(() => javascriptPromise);

		expect(screen.queryByText('Jordan Walke')).toBeNull();
		expect(
			screen.queryByText('Dan Ambramov, Andrew Clark')
			).toBeNull();
		expect(screen.queryByText('Brendan Eich')).toBeInTheDocument();

	});

	// This test is creating warnings
	test('renders snapshot of App', () => {
		const { container } = render(<App />);
		expect(container.firstChild).toMatchSnapshot();
	});

});

