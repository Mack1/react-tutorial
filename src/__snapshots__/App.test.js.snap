// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders snapshot of App 1`] = `
<div>
  <h1>
    My Hacker Stories
  </h1>
  <form>
    <label
      for="search"
    >
      <strong>
        Search:
      </strong>
    </label>
     
    <input
      id="search"
      type="text"
      value="JavaScript"
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
  <hr />
  <p>
    Loading ...
  </p>
</div>
`;

exports[`List renders snapshot of List 1`] = `
<div>
  <span>
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span>
    Jordan Walke
  </span>
  <span>
    3
  </span>
  <span>
    4
  </span>
  <span>
    <button
      type="button"
    >
      Dismiss
    </button>
  </span>
</div>
`;

exports[`SearchForm renders snapshot of SearchForm 1`] = `
<form>
  <label
    for="search"
  >
    <strong>
      Search:
    </strong>
  </label>
   
  <input
    id="search"
    type="text"
    value="React"
  />
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;
